File: teds/lib/gauss_newton.py
# This source code is licensed under the 3-clause BSD license found in
# the LICENSE file in the root directory of this project.
"""Gauss-Newton optimization."""
import numpy as np
import numpy.typing as npt

from .convolution import Kernel
from .radiative_transfer import OpticAbsProp
from .radiative_transfer import nonscat_fwd_model
from .surface import Surface
from teds.l1l2.types import L2
from teds.l1l2.types import RefProfiles
from teds.sgm.atmosphere import Atmosphere


def gauss_newton(config: dict,
                 ref_profiles: RefProfiles,
                 atm: Atmosphere,
                 optics: OpticAbsProp,
                 wave_lbl: npt.NDArray[np.float64],
                 sun_lbl: npt.NDArray[np.float64],
                 sun: npt.NDArray[np.floating],
                 ymeas: npt.NDArray[np.floating],
                 Smeas_diag: npt.NDArray[np.floating],
                 mu0: float,
                 muv: float,
                 isrf: Kernel,
                 timings: dict,
                 l2: L2,
                 i_alt: int,
                 i_act: int) -> None:
    """Nonlinear least squares fit using Gauss-Newton.

    Parameters
    ----------
    retrieval_init
        Initialization dictionary of the least squares fit
    atm
        Atmosphere properties, including trace gas concentrations
    optics
        Optics properties for computing the optical depths
    wave_lbl
        Line-by-line wavelength grid
    sun_lbl
        Solar irradiance on the line-by-line grid
    ymeas
        Radiance spectrum of a given ALT/ACT point
    Smeas_diag
        Radiance spectrum variances
    mu0
        cos(SZA)
    muv
        cos(VZA)
    isrf
        ISRF
    timings
        Collections of measured times. Helps to identify potential
        bottlenecks.
    l2
        Level 2 product (output)
    i_alt
        L2 ALT bin populated with data on this run
    i_act
        L2 ACT bin populated with data on this run

    """
    # Gases to be retrieved
    gas_names = config['retrieval']['gases']
    n_gas = len(gas_names)
    state_vector_names = {}
    for i, gas_name in enumerate(gas_names):
        state_vector_names[gas_name] = i
    n_state = n_gas

    n_albedo = config['retrieval']['n_albedos']
    for i in range(n_albedo):
        state_vector_names['albedo_' + str(i)] = i
    n_state += n_albedo

    if config['retrieval']['do_shift']:
        state_vector_names['shift'] = n_state
        n_state += 1

    # State vector, to be populated with gas concentrations and albedo
    # coefficients.
    state_vector = np.empty(n_state)

    # Start by adding gas scaling initial guesses to the state vector
    for i, gas in enumerate(gas_names):
        xcol = sum(atm.get_gas(gas).concentration) / sum(atm.air)
        # Prior scaling parameter
        state_vector[i] = ref_profiles.initial[gas] / xcol

    # Next add albedo coefficient initial guesses. The first entries
    # are always gases. Thus, slicing with [n_gas:n_gas+n_albedo]
    # allows to access the albedo coefficients.
    state_vector[n_gas:n_gas+n_albedo] = 0
    # Derive first guess albedo from maximum reflectance
    idx = np.argmax(ymeas)
    state_vector[n_gas] = ymeas[idx] / sun[idx] * np.pi / mu0

    if config['retrieval']['do_shift']:
        state_vector[-1] = 0

    chi2 = []
    n_dof = len(ymeas) - len(state_vector)
    surface = Surface(wave_lbl)
    for iterations in range(config['retrieval']['max_iter']):
        # Update all fitted variables using the current state vector
        for i_gas, gas in enumerate(gas_names):
            atm.get_gas(gas).concentration[:] = (
                state_vector[i_gas] * ref_profiles.gases[gas])
        surface.get_albedo_poly(list(state_vector[n_gas:n_gas+n_albedo]))
        if config['retrieval']['do_shift']:
            isrf.regenerate(state_vector[n_gas+n_albedo])

        # Nonscattering forward model
        fwd_rad, derivatives, derivatives_layers = nonscat_fwd_model(
            gas_names,
            n_albedo,
            config['retrieval']['do_shift'],
            isrf,
            sun_lbl,
            atm,
            optics,
            surface,
            mu0,
            muv,
            timings)

        ytilde = ymeas - fwd_rad

        # Calculate Jacobian from convolved quantities
        jacobian = np.empty([len(fwd_rad), len(state_vector)])
        for i, derivative in enumerate(derivatives):
            jacobian[:, i] = derivative

        # Calculate least square solution
        Syinv = np.eye(fwd_rad.size) / Smeas_diag
        JTJ = np.matmul(jacobian.T, np.matmul(Syinv, jacobian))
        Sx = np.linalg.solve(JTJ, np.eye(JTJ.shape[0]))
        gain = np.matmul(Sx, np.matmul(jacobian.T, Syinv))
        state_delta = np.matmul(gain, ytilde)

        precisions = np.empty(len(state_delta))
        for m in range(len(state_delta)):
            precisions[m] = np.sqrt(Sx[m, m])

        state_vector += state_delta

        chi2.append(((ymeas - fwd_rad)**2 / Smeas_diag).sum() / n_dof)

        if iterations > 2 and (chi2[-2] - chi2[-1]
                               < config['retrieval']['chi2_lim']):
            l2.converged[i_alt, i_act] = True
            break

    l2.chi2[i_alt, i_act] = chi2[-1]
    l2.iterations[i_alt, i_act] = iterations + 1

    # Define output product, first update all parameters
    for i_gas, gas in enumerate(gas_names):
        atm.get_gas(gas).concentration[:] = (
            state_vector[i_gas] * ref_profiles.gases[gas])

    # Calculate column mixing ratio, precision, and albedo values
    for i_gas, gas in enumerate(gas_names):
        if gas == 'CH4':
            scaling = 1e-9
        else:
            scaling = 1e-6
        mixing_ratio = (
            sum(atm.get_gas(gas).concentration)/sum(atm.air))
        l2.mixing_ratios[gas][i_alt, i_act] = mixing_ratio
        l2.precisions[gas][i_alt, i_act] = mixing_ratio * precisions[i_gas]
        ref_mixing_ratio = (state_vector[i_gas]
                            * np.sum(ref_profiles.gases[gas])
                            / sum(atm.air) / scaling)
        l2.gains[gas][i_alt, i_act, :] = gain[i_gas] * ref_mixing_ratio
    l2.albedo0[i_alt, i_act] = state_vector[n_gas]

    # Spectral shift
    if config['retrieval']['do_shift']:
        l2.spec_shift[i_alt, i_act] = state_vector[n_gas + n_albedo]
    else:
        l2.spec_shift[i_alt, i_act] = 0

    # Calculate column averaging kernels
    n_lay = atm.zlay.size
    for i_gas, gas in enumerate(gas_names):
        col = np.sum(ref_profiles.gases[gas])
        for k_lay in range(n_lay):
            delta_prof = col / ref_profiles.gases[gas][k_lay]
            l2.col_avg_kernels[gas][i_alt, i_act, k_lay] = (
                np.dot(gain[i_gas, :], derivatives_layers[i_gas][:, k_lay])
                * delta_prof)

File: cpp/l1l2/gauss_newton.cpp
// This source code is licensed under the 3-clause BSD license found
// in the LICENSE file in the root directory of this project.

#include "gauss_newton.h"

#include "atmosphere.h"
#include "nonscat_fwd_model.h"
#include "optic_abs_prop.h"
#include "ref_profiles.h"
#include "settings_l2.h"
#include "surface.h"

#include <common/isrf.h>
#include <common/l2.h>
#include <numeric>

namespace tango {

auto gaussNewton(const int i_gp,
                 const SettingsL2& settings,
                 const RefProfiles& ref_profiles,
                 const Atmosphere& atm_orig,
                 const Eigen::ArrayXd& wave_lbl,
                 const Eigen::ArrayXd& sun_lbl,
                 const Eigen::Ref<const Eigen::ArrayXd> spectrum,
                 const Eigen::Ref<const Eigen::ArrayXd> S_meas,
                 const double mu0,
                 const double muv,
                 const ISRF& isrf,
                 OpticAbsProp& optics,
                 const Eigen::ArrayXd& sun,
                 L2& l2) -> void
{
    // A copy of the standard atmosphere is use for storing the
    // retrieved gas columns.
    Atmosphere atm { atm_orig };

    // Gases to be retrieved
    std::vector<std::string> gas_names {};
    for (auto it { ref_profiles.gases.begin() }; it != ref_profiles.gases.end();
         ++it) {
        gas_names.push_back(it->first);
    }
    const int n_gases { static_cast<int>(gas_names.size()) };
    const int n_albedos { settings.retrieval.n_albedos };

    // State vector, to be populated with gas concentrations and
    // albedo coefficients.
    Eigen::VectorXd state_vector(n_gases + n_albedos);

    // Start by adding gas scaling initial guesses to the state vector
    for (int i {}; i < n_gases; ++i) {
        state_vector(i) =
          ref_profiles.initial.at(gas_names[i])
          / (ref_profiles.gas_sums.at(gas_names[i]) / atm.air_sum);
    }

    // Next add albedo coefficient initial guesses. The first entries
    // of the state vector correspond to gases.
    state_vector(Eigen::seqN(n_gases, n_albedos)).array() = 0.0;
    // Derive first guess albedo from maximum reflectance
    Eigen::Index idx_max {};
    const double max_rad { spectrum.maxCoeff(&idx_max) };
    state_vector(n_gases) = max_rad / sun(idx_max) * std::numbers::pi / mu0;

    // Initialize Gauss-Newton work variables
    const int n_dof { static_cast<int>(spectrum.size() - state_vector.size()) };
    double chi2_prev { 1e10 };
    Surface surface { n_albedos, wave_lbl };
    Eigen::VectorXd rad(spectrum.size());
    Eigen::MatrixXd jacobian(spectrum.size(), state_vector.size());
    Eigen::ArrayXd rad_lbl {};
    Eigen::ArrayXd dev_tau_lbl {};
    Eigen::ArrayXd dev_alb_lbl {};
    Eigen::ArrayXd precisions {};
    Eigen::MatrixXd gains {};

    for (int iterations {}; iterations < settings.retrieval.max_iter;
         ++iterations) {
        for (int i_gas {}; i_gas < n_gases; ++i_gas) {
            atm.gases.at(gas_names[i_gas]) =
              state_vector[i_gas] * ref_profiles.gases.at(gas_names[i_gas]);
        }

        surface.getAlbedoPoly(state_vector(Eigen::seqN(n_gases, n_albedos)));

        nonscatFwdModel(state_vector,
                        gas_names,
                        n_albedos,
                        isrf,
                        sun_lbl,
                        atm,
                        surface,
                        mu0,
                        muv,
                        optics,
                        rad_lbl,
                        dev_tau_lbl,
                        dev_alb_lbl,
                        rad,
                        jacobian);

        const Eigen::VectorXd y_diff = spectrum.matrix() - rad;

        // Calculate least square solution
        const Eigen::MatrixXd Syinv = (1.0 / S_meas).matrix().asDiagonal();
        const Eigen::MatrixXd JTJ = jacobian.transpose() * (Syinv * jacobian);
        const Eigen::MatrixXd Sx =
          JTJ.llt().solve(Eigen::MatrixXd::Identity(JTJ.rows(), JTJ.cols()));
        gains = Sx * (jacobian.transpose() * Syinv);
        const Eigen::VectorXd state_delta = gains * y_diff;

        state_vector += state_delta;

        l2.chi2(i_gp) = (y_diff.array().square() / S_meas).sum() / n_dof;
        l2.iterations(i_gp) = iterations + 1;

        const bool converged { chi2_prev - l2.chi2(i_gp)
                               < settings.retrieval.chi2_lim };
        if (converged || iterations == settings.retrieval.max_iter - 1) {
            precisions = Sx.diagonal().array().sqrt();
        }
        if (converged) {
            l2.converged(i_gp) = true;
            break;
        }
        chi2_prev = l2.chi2(i_gp);
    }

    // Update gas concentrations
    for (int i_gas {}; i_gas < n_gases; ++i_gas) {
        atm.gases.at(gas_names[i_gas]) =
          state_vector[i_gas] * ref_profiles.gases.at(gas_names[i_gas]);
    }

    // Column mixing ratios, precisions, and albedo values
    for (int i_gas {}; i_gas < n_gases; ++i_gas) {
        const std::string& gas { gas_names[i_gas] };
        l2.mixing_ratios.at(gas)(i_gp) = atm.gases.at(gas).sum() / atm.air_sum;
        l2.precisions.at(gas)(i_gp) =
          l2.mixing_ratios.at(gas)(i_gp) * precisions[i_gas];
        l2.gains.at(gas).row(i_gp) = gains.row(i_gas)
                                     * l2.mixing_ratios.at(gas)(i_gp)
                                     * (gas == "CH4" ? 1e9 : 1e6);
    }
    l2.albedo0(i_gp) = state_vector(n_gases);

    // Column averaging kernels
    optics.setOptDepthLayers(atm.gases, gas_names);
    Eigen::MatrixXd deriv_gas_layers(spectrum.size(), atm.z_lay.size());
    for (int i_gas {}; i_gas < n_gases; ++i_gas) {
        const std::string& gas { gas_names[i_gas] };
        speciesDerivLayers(
          gas, isrf, optics.tau_alt.at(gas), dev_tau_lbl, deriv_gas_layers);
        l2.col_avg_kernels.at(gas).row(i_gp) =
          (gains.row(i_gas) * deriv_gas_layers).array()
          * ref_profiles.gas_sums.at(gas)
          / ref_profiles.gases.at(gas).transpose();
    }
}

} // namespace tango
